src/rust/inetstack/protocols/layer4/tcp/established/sender.rs

```diff
@@ -459,18 +459,23 @@ impl Sender {
             // Check and update send window if necessary.
             self.update_send_window(header);
 
-            if header.ack_num == send_next {
-                // This segment acknowledges everything we've sent so far (i.e. nothing is currently outstanding).
-                // Since we no longer have anything outstanding, we can turn off the retransmit timer.
-                debug_assert_eq!(self.unacked_queue.is_empty(), true);
-                self.retransmit_deadline_time_secs.set(None);
-            } else {
-                // Update the retransmit timer.  Some of our outstanding data is now acknowledged, but not all.
-                // TODO: This looks wrong.  We should reset the retransmit timer to match the deadline for the
-                // oldest still-outstanding data.  The below is overly generous (minor efficiency issue).
-                let deadline: Instant = now + self.rto_calculator.rto();
-                self.retransmit_deadline_time_secs.set(Some(deadline));
-            }
+            // Reset the retransmit timer if necessary. If there is more data that hasn't been acked, then set to the
+            // next segment deadline, otherwise, do not set.
+            let retransmit_deadline_time_secs: Option<Instant> = match self.unacked_queue.get_front() {
```

Comment:
The logic looks good. This is a good candidate for extracting into a separate function as it represents a nice logically independent block.